### DISCUSSION TITLE
LR1110MB1XXS with STM32L476, sleep mode settings

### DISCUSSION THREAD

**Discussion by sohaibqamar** [Lora-net/LoRaMac-node#1572] [Q&A]
Hello everyone, 
I am using a custom board for a LoRaWAN-based solution and using this stack for the development. 
I'm using the below settings.
```
"APPLICATION":"LoRaMac",
"SUB_PROJECT":"periodic-uplink-lpp",
"LORAWAN_DEFAULT_CLASS":"CLASS_A",
"CLASSB_ENABLED":"ON",
"MODULATION":"LORA",
"BOARD":"NucleoL476",
"MBED_RADIO_SHIELD":"LR1110MB1XXS",
"SECURE_ELEMENT":"SOFT_SE"
```
I have included a GPS over UART in the project and used some GPIOs for LEDs and Buzzer. I am using AWS IoT Core for Uplink/Downlinks which is working perfectly fine. Periodic time for sending uplink is 1min.
My system is not going into sleep mode or any low-power mode using the default while(1) loop.
I am measuring the current using a power profiler and the system current is around 50-60mA, I am unable to go into any sleep mode.
Any help would be appreciated.

**Comment by sohaibqamar** [comment]
My main function looks like this after I deleted the portion of interfacing the Accelerometer and GPS

```
int main( void )
{
    BoardInitMcu( );
    BoardInitPeriph( );
    TimerInit( &Led1Timer, OnLed1TimerEvent );
    TimerSetValue( &Led1Timer, 25 );
    TimerInit( &Led2Timer, OnLed2TimerEvent );
    TimerSetValue( &Led2Timer, 25 );
    TimerInit( &LedBeaconTimer, OnLedBeaconTimerEvent ); // What is this timer doing
    // TimerSetValue( &LedBeaconTimer, 5000 ); // original
    TimerSetValue( &LedBeaconTimer, 30000 );
    // Initialize transmission periodicity variable
    TxPeriodicity = APP_TX_DUTYCYCLE + randr( -APP_TX_DUTYCYCLE_RND, APP_TX_DUTYCYCLE_RND );
    const Version_t appVersion = { .Value = FIRMWARE_VERSION };
    const Version_t gitHubVersion = { .Value = GITHUB_VERSION };
    DisplayAppInfo( "periodic-uplink-lpp", 
                    &appVersion,
                    &gitHubVersion );    
    if ( LmHandlerInit( &LmHandlerCallbacks, &LmHandlerParams ) != LORAMAC_HANDLER_SUCCESS )
    {
        printf( "LoRaMac wasn't properly initialized\n" );
        // Fatal error, endless loop.
        while ( 1 )
        {
            printf( "* " );
        }
    }
    // Set system maximum tolerated rx error in milliseconds
    LmHandlerSetSystemMaxRxError( 20 );
    // The LoRa-Alliance Compliance protocol package should always be
    // initialized and activated.
    LmHandlerPackageRegister( PACKAGE_ID_COMPLIANCE, &LmhpComplianceParams );
    LmHandlerJoin( );
    StartTxProcess( LORAMAC_HANDLER_TX_ON_TIMER );    
    while( 1 )
    {
        // Processes the LoRaMac events
        LmHandlerProcess( );
        // Process application uplinks management
        UplinkProcess( );
        CRITICAL_SECTION_BEGIN( );
        if( IsMacProcessPending == 1 )
        {
            // Clear flag and prevent MCU to go into low power modes.
            IsMacProcessPending = 0;
        }
        else
        {
            // The MCU wakes up through events
            BoardLowPowerHandler( );
            //printf("Wake up.\n");
        }
        CRITICAL_SECTION_END( );
    }
}
```
In function `void BoardInitMcu( void )` i have set `UsbIsConnected = false;`

**Comment by mluis1** [comment]
Is is hard to help you figure out why such high power consumption is observed.

There are multiple reasons to have such behaviors.
As you indicate that you have your own custom board it is even harder to help you as we do not know how your board has been designed.

The only thing we can do is to provide some clues on what you should look at.
We suppose that `BoardLowPowerHandler` function gets called. This function execution is where the MCU enters the sleep modes.

Things to verify related to your custom board:
- Are there any LED that is ON?
- Check that MCU pins are not floating or are set to the right level when in sleep
- Check that there is no debug probe connected to your custom board
- Are external peripheral component set in sleep mode as well (temperature sensors, GPS receivers, etc)
- etc

Debugging why one observes a higher power consumption than expected is a hard task and very time consuming.

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by Yousaf93** [reply]
Hi m I hope you are good thanks for the response.
The board is fairly simple and follows all the same conneectivity as a nucleo476 baord with LR1110 does. Only additional thing would be the GPS which is also not configured now to test the normal operations and sleeps, 
I am going to re verify your suggestions and see if we are missing anything and get back to you shortly

**Reply by Yousaf93** [reply]
In the periodic uplink example arent all pins already set to correct modes before sleeping ? just asking because we didnt modify that setup

**Reply by mluis1** [reply]
The pins are set in the correct modes for the NucleoL476+LR1110 board combination.

However, your custom board may use other pins or peripherals. Therefore you need to take care of them.

**Reply by Yousaf93** [reply]
i am not initialization any other pin or there modes but yes one of the controller pin is connected to PPS pin of GPS , and the pin is not configuired any where in my code , would that be an issue ? I thought if i configure then i need to handle it

**Reply by Yousaf93** [reply]
As you mention debug prob are you refering to a USB to serial converter ? Becaues i use one to see debug prints , and also when i supply power from my power profiler the SWDIO and SWCLK are still connected to STLink ,
Please guide me these points i am in testing right now have been trying around this from couple of weeks now and its the last hurdle before compeleting my work

**Reply by mluis1** [reply]
The ST-Link is the debug probe.

I just remembered that for Nucleo and B-L072Z-LRWAN1 platforms the MCU debug functionality is always ON.

Using the MCU debug peripheral does add power consumption when in one of the low power modes.
Please try to change `BoardUnusedIoInit` function from:
https://github.com/Lora-net/LoRaMac-node/blob/6384af8f4c4a57016c83f3d34430f0f68855e695/src/boards/NucleoL476/board.c#L346-L351

to something similar to:
```c
static void BoardUnusedIoInit( void )
{
#if defined( USE_DEBUGGER )
    HAL_DBGMCU_EnableDBGSleepMode( );
    HAL_DBGMCU_EnableDBGStopMode( );
    HAL_DBGMCU_EnableDBGStandbyMode( );
#else
    HAL_DBGMCU_DisableDBGSleepMode( );
    HAL_DBGMCU_DisableDBGStopMode( );
    HAL_DBGMCU_DisableDBGStandbyMode( );
#endif
}
```

Like this the debug functionality under low power modes is only possible if `USE_DEBUGGER` is defined at compile time.

While performing the power consumption measurements I would also advise to remove the USB to Serial converter connection.

**Reply by Yousaf93** [reply]
Hi, Thanks i will check with @sohaibqamar 
As you mentioned high current in low power modes but in our case we are not able to go to low power mode

**Reply by Yousaf93** [reply]
Does "USE_DEBUGGER" needs to be turned off in Cmakelists  or development-environment.md

**Comment by sohaibqamar** [comment]
@mluis1 I am still unable to figure out the problem, now I am only testing the sleep functionality with the LoRaWAN code all commented
Now my `main `function looks like the below
```
int main( void )
{
    BoardInitMcu( );
    BoardInitPeriph( );
    
    while( 1 )
    {
        GpioWrite( &Led1, 1 );
        HAL_Delay(1000);
        GpioWrite( &Led1, 0 );
        HAL_Delay(1000);

        CRITICAL_SECTION_BEGIN( );
        BoardLowPowerHandler( );
        CRITICAL_SECTION_END( );
    }
}
```
My BoardInitMcu function looks like the below
```
void BoardInitMcu( void )
{
    if( McuInitialized == false )
    {
        HAL_Init( );

        // InitFlashMemoryOperations( );

        // LEDs
        GpioInit( &Led1, LED_1, PIN_OUTPUT, PIN_PUSH_PULL, PIN_NO_PULL, 0 );

        SystemClockConfig( );

        UsbIsConnected = false; // true;

        RtcInit( );

        BoardUnusedIoInit( );        
        
        if( GetBoardPowerSource( ) == BATTERY_POWER )
        {
            // Disables OFF mode - Enables lowest power mode (STOP)
            LpmSetOffMode( LPM_APPLI_ID, LPM_DISABLE );
        }
    }
    else
    {
        SystemClockReConfig( );
    }

#if defined( SX1261MBXBAS ) || defined( SX1262MBXCAS ) || defined( SX1262MBXDAS )
    SpiInit( &SX126x.Spi, SPI_1, RADIO_MOSI, RADIO_MISO, RADIO_SCLK, NC );
    SX126xIoInit( );
#elif defined( LR1110MB1XXS )
    // SpiInit( &LR1110.spi, SPI_1, RADIO_MOSI, RADIO_MISO, RADIO_SCLK, NC );
    // lr1110_board_init_io( &LR1110 );
#elif defined( SX1272MB2DAS )
    SpiInit( &SX1272.Spi, SPI_1, RADIO_MOSI, RADIO_MISO, RADIO_SCLK, NC );
    SX1272IoInit( );
#elif defined( SX1276MB1LAS ) || defined( SX1276MB1MAS )
    SpiInit( &SX1276.Spi, SPI_1, RADIO_MOSI, RADIO_MISO, RADIO_SCLK, NC );
    SX1276IoInit( );
#endif

    if( McuInitialized == false )
    {
        McuInitialized = true;
#if defined( SX1261MBXBAS ) || defined( SX1262MBXCAS ) || defined( SX1262MBXDAS )
        SX126xIoDbgInit( );
        // WARNING: If necessary the TCXO control is initialized by SX126xInit function.
#elif defined( LR1110MB1XXS )
        // lr1110_board_init_dbg_io( &LR1110 );
        // WARNING: If necessary the TCXO control is initialized by SX126xInit function.
#elif defined( SX1272MB2DAS )
        SX1272IoDbgInit( );
        SX1272IoTcxoInit( );
#elif defined( SX1276MB1LAS ) || defined( SX1276MB1MAS )
        SX1276IoDbgInit( );
        SX1276IoTcxoInit( );
#endif
    }
}
```
The program blinks LED once and goes to sleep forever, I want to know which part of the code is handling the wakeup from sleep using RTC. Is there any relation of macro `#define APP_TX_DUTYCYCLE 5000` with the sleep time?
It would be of great help, please.
Sohaib
